### PR TITLE
Add New Resource azuredevops_serviceendpoint_github_enterprise

### PR DESCRIPTION
**Pull request by uanid**
## All Submissions:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission.
* [x] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?

Add New Resource `azuredevops_serviceendpoint_github_enterprise`
Issue Number: #331 #210 #28

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [ ] Yes
- [x] No

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Any relevant logs, error output, etc?

(If it’s long, please paste to https://ghostbin.com/ and insert the link here.)

## Other information

New Resource
```terraform
resource "azuredevops_serviceendpoint_github_enterprise" "gh_serviceendpoint" {
  project_id            = azuredevops_project.project.id
  service_endpoint_name = "GitHub Service Connection"

  url  = "https://github.contoso.com"

  auth_personal {
    # personal_access_token = "..." Or set with `AZDO_GITHUB_ENTERPRISE_SERVICE_CONNECTION_PAT` env var
  }
}
```

Most of attributes are same with `azuredevops_serviceendpoint_github`
Only added `url` attribute.
Users can add url attributes to your self-hosted Github-Etnerprise-Server url.

### CODE REVIEW

**Comment by xuzhang3** [comment]
Hi @Uanid Suggest create a new resource `azuredevops_serviceendpoint_github_enterprise`.  Add new configurations to `azuredevops_serviceendpoint_github` is not the best choice, some of the configurations only apply to the corresponding resources.

**Comment by uanid** [comment]
@xuzhang3 Ok. I will change it.

**Comment by uanid** [comment]
@xuzhang3 
Hello.
I added new resource `azuredevops_serviceendpoint_github_enterprise` and updated docs.
Can you re-check this PullRequest?

File: .azdo/main.tf [hunk 1]
```diff
@@ -20,8 +20,11 @@ resource "azuredevops_serviceendpoint_github" "github_serviceendpoint" {
   project_id            = azuredevops_project.project.id
   service_endpoint_name = "GitHub Service Connection"
 
+  type = "github"
+  url  = "https://github.com"
+
```
Comment (xuzhang3) [inline .azdo/main.tf]: These configurations should be removed.

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github.go [hunk 2]
```diff
@@ -118,12 +119,15 @@ func flattenServiceEndpointGitHub(d *schema.ResourceData, serviceEndpoint *servi
 			d.Set("auth_personal", authPersonal)
 		}
 	}
+
+	d.Set("type", *serviceEndpoint.Type)
+	d.Set("url", *serviceEndpoint.Url)
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github.go]: Should be removed

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github_enterprise.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+	if strings.EqualFold(*serviceEndpoint.Authorization.Scheme, "Token") {
+		authPersonalSet := d.Get("auth_personal").(*schema.Set).List()
+		authPersonal := flattenAuthPersonGithubEnterprise(d, authPersonalSet)
+		if authPersonal != nil {
+			d.Set("auth_personal", authPersonal)
+		}
+	}
+
+	d.Set("type", *serviceEndpoint.Type)
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github_enterprise.go]: Should be removed,  this setting doesn't exist.

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github_test.go [hunk 4]
```diff
@@ -149,10 +149,10 @@ func TestServiceEndpointGitHub_Update_DoesNotSwallowError(t *testing.T) {
 	require.Contains(t, err.Error(), "UpdateServiceEndpoint() Failed")
 }
 
-func configureAuthPersonal(d *schema.ResourceData) {
+func configureGhAuthPersonal(d *schema.ResourceData) {
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github_test.go]: `configureGhAuthPersonal` should be rename to `configureAuthPersonal`, this tests are for `resource_serviceendpoint_github`

File: website/docs/r/serviceendpoint_github_enterprise.html.markdown [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  visibility         = "private"
+  version_control    = "Git"
+  work_item_template = "Agile"
+}
+
+resource "azuredevops_serviceendpoint_github_enterprise" "serviceendpoint_ghes_1" {
+  project_id            = azuredevops_project.project.id
+  service_endpoint_name = "GHES"
+  url                   = "https://github.contoso.com"
+  description           = "Sample GithHub Enterprise Personal Access Token"
```
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_github_enterprise.html.markdown]:
```suggestion
  service_endpoint_name = "Sample GithHub Enterprise"
  url                   = "https://github.contoso.com"
  description           = "Managed by Terraform""
```

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github_enterprise.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+	scheme := "InstallationToken"
+	parameters := map[string]string{}
+
+	if config, ok := d.GetOk("auth_personal"); ok {
+		scheme = "Token"
+		parameters = expandAuthPersonalSetGithubEnterprise(config.(*schema.Set))
+	}
+
+	if config, ok := d.GetOk("auth_oauth"); ok {
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github_enterprise.go]: `auth_oauth`  is not  supported by by the schema.  If you want support `auth_oauth`, you need add the `auth_oauth` to schema like `auth_personal` does

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github_enterprise.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+		},
+	}
+	patHashKey, patHashSchema := tfhelper.GenerateSecreteMemoSchema(personalAccessTokenGithubEnterprise)
+	authPersonal.Schema[patHashKey] = patHashSchema
+	r.Schema["auth_personal"] = &schema.Schema{
+		Type:     schema.TypeSet,
+		Optional: true,
+		MinItems: 1,
+		MaxItems: 1,
+		Elem:     authPersonal,
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github_enterprise.go:38]: If want support both personal access token and Oauth. I would suggest add a constraint like `ExactlyOneOf: []string{"auth_personal", "auth_oauth"},` for both `auth_personal` and `auth_oauth`
If only support `auth_personal`,  you should change `Optional: true,` to `Required: true,`

**Review by xuzhang3 — CHANGES_REQUESTED** [review]
Hi @Uanid thanks for your contribution, a few changes requested.
1. add `serviceendpoint_github_enterprise.html.markdown` to `azuredevops.erb`, otherwise
2. add acceptance tests for  `azuredevops_serviceendpoint_github_enterprise` like [serviceendpoint_github](https://github.com/microsoft/terraform-provider-azuredevops/blob/master/azuredevops/internal/acceptancetests/resource_serviceendpoint_github_test.go)

File: .azdo/main.tf (hunk 1 above)
Comment (uanid) [reply .azdo/main.tf]: Deleted configurations

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github.go (hunk 2 above)
Comment (uanid) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github.go]: Delete configurations

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github_test.go (hunk 4 above)
Comment (uanid) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github_test.go]: Renamed function

File: website/docs/r/serviceendpoint_github_enterprise.html.markdown (hunk 5 above)
Comment (uanid) [reply website/docs/r/serviceendpoint_github_enterprise.html.markdown]: Changed attribute values

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github_enterprise.go (hunk 6 above)
Comment (uanid) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github_enterprise.go]: Deleted auth_oauth codes

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github_enterprise.go (hunk 7 above)
Comment (uanid) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github_enterprise.go:38]: Deleted Optional attribute & Add required attribute

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github_enterprise.go (hunk 3 above)
Comment (Uanid) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github_enterprise.go:63]: When I use `az devops service-endpoint show --id=<service-endpoint-id>` and returnd type to `githubenterprise`
In my think, AZDO server has return `type` attributes and if I use `terraform import` then this code is needed.
Should I remove this codes?

Comment (uanid) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_github_enterprise.go]: Deleted settings

**Review by xuzhang3 — CHANGES_REQUESTED** [review]
Hi @Uanid  you need add acceptance tests for `azuredevops_serviceendpoint_github_enterprise`. 
Reference:
1.  [Acceptance Tests](https://www.terraform.io/docs/extend/testing/acceptance-tests/index.html)
2. [azuredevops_serviceendpoint_github](https://github.com/microsoft/terraform-provider-azuredevops/blob/master/azuredevops/internal/acceptancetests/resource_serviceendpoint_github_test.go)

**Comment by uanid** [comment]
Hi @xuzhang3 
Thank you for support commits this PR.
I added ACC TEST for Github Enterprise.
And tested in my local PC with below env vars.
스크린샷 2021-04-08 오후 7 18 14

File: azuredevops/internal/acceptancetests/resource_serviceendpoint_github_enterprise_test.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+)
+
+func TestAccServiceEndpointGitHubEnterprise_PersonalTokenBasic(t *testing.T) {
+	projectName := testutils.GenerateResourceName()
+	serviceEndpointName := testutils.GenerateResourceName()
+
+	resourceType := "azuredevops_serviceendpoint_github_enterprise"
+	tfSvcEpNode := resourceType + ".serviceendpoint"
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testutils.PreCheck(t, &[]string{"AZDO_GITHUB_ENTERPRISE_SERVICE_CONNECTION_PAT"}) },
```
Comment (xuzhang3) [inline azuredevops/internal/acceptancetests/resource_serviceendpoint_github_enterprise_test.go]: `personal_access_token` can be set with dummy values,  remove `AZDO_GITHUB_ENTERPRISE_SERVICE_CONNECTION_PAT` from the the precheck,  so we do not need to update the CI pipeline when new environment variables add to the provider.

**Review by xuzhang3 — CHANGES_REQUESTED** [review]

**Comment by xuzhang3** [comment]
you can run the ACCTESTs in terminal with `make testacc TESTARGS='-run=TestAccServiceEndpointGitHubEnterprise' TESTTAGS='all'`, make sure you have `make` installed and set the environment 
```shell
export AZDO_ORG_SERVICE_URL=https://dev.azure.com/xxxx/
export AZDO_PERSONAL_ACCESS_TOKEN=xxxx
```

**Comment by uanid** [comment]
Thanks for acctest tips!
Could you re-review TestAccServiceEndpointGitHubEnterprise?

**Comment by xuzhang3** [comment]
LGTM

**Comment by uanid** [comment]
Thank you for support!
I really need this features.

**Comment by xuzhang3** [comment]
> Thank you for support!
> I really need this features.

you can build a private package with `go build -gcflags="all=-N -l" -o terraform-provider-azuredevops` before v0.1.4 released.

**Comment by uanid** [comment]
> > Thank you for support!
> > I really need this features.
> 
> you can build a private package with `go build -gcflags="all=-N -l" -o terraform-provider-azuredevops` before v0.1.4 released.

스크린샷 2021-04-09 오후 3 22 45


I'm already use how you says. :)

**Pull request merged**